### PR TITLE
Fix raster masks access from above the pixelpipe

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
While checking for an available rastermask we look through all modules in the pipe for a "hit for candidate".

This is fine as long as we remember the origin of the raster masks and avoid wrong "reordering of the pipe".

What we miss for safety is a test for
- we haven't found the source module for the mask until the target has been reached
- some iop-order issue

Fixes: #13276

Release note: Fix crashes while using raster masks after reordering the pixelpipe

### CODE REVIEW

File: src/develop/pixelpipe_hb.c [hunk 1]
```diff
@@ -2880,9 +2880,21 @@ float *dt_dev_get_raster_mask(const dt_dev_pixelpipe_t *pipe,
...
-  for(source_iter = pipe->nodes; source_iter; source_iter = g_list_next(source_iter))
+  for(source_iter = piece->pipe->nodes; source_iter; source_iter = g_list_next(source_iter))
   {
     const dt_dev_pixelpipe_iop_t *candidate = (dt_dev_pixelpipe_iop_t *)source_iter->data;
+
+    if((candidate->module == target_module)
+       || (candidate->module->iop_order >= target_module->iop_order))
+    {
+      dt_control_log
+        (_("can't take a rastermask from later in the pipeline"));
```
Comment (TurboGit) [inline src/develop/pixelpipe_hb.c:2891]: Would be nice to have more information here, what about:

`module '%s' can't take raster mask from module '%s' which is later in the pixel pipe`

Or something like that.

Comment (TurboGit) [reply src/develop/pixelpipe_hb.c:2891]: And adding also:

`raster mask is ignored`

as this is in fact non fatal and reordering the module back to proper order does restore the raster mask.

Comment (jenshannoschwalm) [reply src/develop/pixelpipe_hb.c]: done and agreed

**Comment by jenshannoschwalm** [comment]
While being here, may i ask to have a look into the `dt_dev_get_raster_mask()` code in detail?

How if free_mask intended to work?
1. we initialize it to FALSE at start
2. at around L. 2940 we possibly free it. (never done because of (1) Do you know if it was intended to replace the found raster by a transformed one ( that wouldn't work)? Or some sort of cleanup (then we would have to remove the raster mask from the source module here?
3. It seems the only purpose here is to return a flag if the returned raster mask is a "transformed duplicate" and thus should be deallocated after use.
4. Why are some image callers passing a TRUE here?

**Comment by TurboGit** [comment]
A big mess...

1. Why not!
2. Indeed never freed, and seems ok to not free it as it is retrieved from a hash-table, I don't see why it should be freed here.
3. Right.
4. For nothing it seems...

The code is original from initial commit: 313a2a034758b1b6b9006cb2a226891a316e5f1d so we cannot say it is a bug introduced after refactoring.

To me this is wrong and possibly the fix is to just remove:

`                if(*free_mask) dt_free_align(raster_mask);`

And indeed when `free_mask` is returned as `TRUE` the caller should free it. That's how I understand this.

**Comment by jenshannoschwalm** [comment]
Ooh, i think our so-far understanding was wrong and leads to mem-leaks. Added comments into the code ...

**Comment by TurboGit** [comment]
> Ooh, i think our so-far understanding was wrong and leads to mem-leaks. Added comments into the code ...

Indeed, we are in a loop.

File: src/develop/pixelpipe_hb.c [hunk 2]
```diff
@@ -2880,13 +2891,30 @@ float *dt_dev_get_raster_mask(const dt_dev_pixelpipe_t *pipe,
...
     const dt_dev_pixelpipe_iop_t *candidate = (dt_dev_pixelpipe_iop_t *)source_iter->data;
+
+    if((candidate->module == target_module)
+       || (candidate->module->iop_order >= target_module->iop_order))
+    {
+      dt_control_log
+        (_("module `%s' can't get raster mask from module `%s'"
+           "as that is processed later in the pixel pipe.\n"
+           "raster mask is ignored."),
+           target_module->so->op, raster_mask_source->so->op);
```
Comment (TurboGit) [inline src/develop/pixelpipe_hb.c:2905]: I'll update this to use the module name and not the file-name. That is, using target_module->name() which is also translated. Also adding a space before 'as'.

**Pull request merged**